python/tskit/trees.py

```diff
@@ -4160,18 +4160,24 @@ def mean_descendants(self, reference_sets, windows=None):
         :return: An array with dimensions (number of nodes in the tree sequence,
             number of reference sets)
         """
-        denom = self.sample_count_stat(
-                    [self.samples()], lambda x: x > 0, output_dim=1,
-                    polarised=True, strict=False, mode="node", span_normalise=False,
-                    windows=windows)
+        def f(x):
+            out = x.copy()
+            out[-1] = (out[-1] > 0)
+            return out
+
+        sample_sets = reference_sets + [self.samples()]
         C = self.sample_count_stat(
-                    reference_sets, lambda x: x, output_dim=len(reference_sets),
+                    sample_sets, f, output_dim=len(sample_sets),
                     polarised=True, strict=False, mode="node", span_normalise=False,
                     windows=windows)
+        out_slice = tuple([slice(None)] * (len(C.shape) - 1) + [slice(0, -1)])
+        denom_slice = tuple([slice(None)] * (len(C.shape) - 1) + [-1])
+        out = C[out_slice]
+        denom = C[denom_slice].reshape(list(C.shape[:-1]) + [1])
```

Comment:
This implementation is more opaque than [the previous implementation](https://github.com/tskit-dev/tskit/pull/328/commits/ab4dbd9aba581637f00c69ed1ded54540a93e065#diff-993cce8b29b806d11aae4e2ead3c71ecR4164) but is closer to how we'd do it in C.